File: backend/app/services/report_service.py
"""Report submission and review logic backed by Firestore.

TODO: wire actual Firestore reads/writes once collections are finalized.
Trust weighting: field_official reports should carry a higher trust_weight
(e.g. 2.0) than citizen reports (e.g. 1.0) when feeding the risk model.
"""

from app.models.user import ReportStatus, UserRole

TRUST_WEIGHTS = {
    UserRole.CITIZEN: 1.0,
    UserRole.FIELD_OFFICIAL: 2.0,
    UserRole.ANALYST_ADMIN: 2.0,
}


class ReportService:
    def create_report(self, reporter_uid: str, reporter_role: UserRole, payload: dict) -> dict:
        # TODO: persist to Firestore "reports" collection with trust_weight and status=PENDING
        raise NotImplementedError

    def list_reports(self, district: str | None = None, status: ReportStatus | None = None) -> list[dict]:
        # TODO: query Firestore "reports" collection with optional filters
        raise NotImplementedError

    def review_report(self, report_id: str, status: ReportStatus, reviewer_notes: str | None) -> dict:
        # TODO: update Firestore doc; analyst/admin only
        raise NotImplementedError


report_service = ReportService()

File: app/lib/core/services/report_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../../models/report.dart';
import '../constants/app_constants.dart';

/// Reads/writes reports directly against Firestore (simpler for a hackathon
/// prototype than round-tripping through the backend for CRUD; the backend
/// report endpoints remain useful for server-side validation/trust-weight
/// logic once that's implemented).
class ReportService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  Stream<List<Report>> watchReports({String? district}) {
    Query<Map<String, dynamic>> query = _firestore
        .collection(FirestoreCollections.reports)
        .orderBy('created_at', descending: true);

    if (district != null) {
      query = query.where('district', isEqualTo: district);
    }

    return query.snapshots().map(
          (snapshot) => snapshot.docs
              .map((doc) => Report.fromFirestore(doc.id, doc.data()))
              .toList(),
        );
  }

  /// TODO: upload media_urls via firebase_storage first (see
  /// core/services/storage_service.dart), then write this doc with the
  /// resulting download URLs.
  Future<void> submitReport({
    required String reporterUid,
    required UserRole reporterRole,
    required String description,
    required double lat,
    required double lng,
    required String district,
    List<String> mediaUrls = const [],
    RoadStatus? roadStatus,
  }) {
    throw UnimplementedError('TODO: write report doc to Firestore');
  }

  Future<void> reviewReport(String reportId, ReportStatus status) {
    throw UnimplementedError('TODO: analyst/admin approve or reject a report');
  }
}
